### PR TITLE
[test] Add js-api tests

### PR DESCRIPTION
**Pull request by binji**
This tests the `WebAssembly.Memory` constructor and the `grow` method.

### CODE REVIEW

**Comment by binji** [comment]
I haven't tested this yet against an implementation, so it's probably broken :-}

**Comment by binji** [comment]
cc @Ms2ger (can't add you as a reviewer)

File: test/js-api/memory/constructor.any.js [hunk 1]
```diff
@@ -8,15 +8,18 @@ function assert_Memory(memory, expected) {
 
   // https://github.com/WebAssembly/spec/issues/840
   assert_equals(memory.buffer, memory.buffer, "buffer should be idempotent");
-  assert_equals(Object.getPrototypeOf(memory.buffer), ArrayBuffer.prototype,
-                "prototype of buffer");
-  assert_true(Object.isExtensible(memory.buffer), "buffer extensibility");
+  const isShared = !!expected.shared;
+  const bufferType = isShared ? SharedArrayBuffer : ArrayBuffer;
```
Comment (Ms2ger) [inline test/js-api/memory/constructor.any.js]: I think this is going to cause a `ReferenceError` if `SharedArrayBuffer` is not present, even in the not-shared case. Can you use `self.SharedArrayBuffer` instead?

File: test/js-api/memory/constructor.any.js [hunk 2]
```diff
@@ -8,15 +8,18 @@ function assert_Memory(memory, expected) {
...
+  assert_equals(Object.getPrototypeOf(memory.buffer), bufferType.prototype,
+                'prototype of buffer');
   assert_equals(memory.buffer.byteLength, 0x10000 * expected.size, "size of buffer");
   if (expected.size > 0) {
     const array = new Uint8Array(memory.buffer);
     assert_equals(array[0], 0, "first element of buffer");
     assert_equals(array[array.byteLength - 1], 0, "last element of buffer");
   }
+  assert_equals(isShared, Object.isFrozen(memory.buffer), "buffer frozen");
+  assert_not_equals(isShared, Object.isExtensible(memory.buffer), "buffer extensibility");
```
Comment (Ms2ger) [inline test/js-api/memory/constructor.any.js]: Nit: swap the order of the arguments (actual first, expected second).

File: test/js-api/memory/constructor.any.js [hunk 3]
```diff
@@ -118,13 +125,25 @@ test(() => {
         },
       };
     },
+
+    get shared() {
+      order.push("shared");
+      return {
+        valueOf() {
+          order.push("shared valueOf");
+          return 1;
```
Comment (Ms2ger) [inline test/js-api/memory/constructor.any.js]: Maybe `true`

File: test/js-api/memory/grow.any.js [hunk 4]
```diff
@@ -2,7 +2,8 @@
 
 function assert_ArrayBuffer(actual, expected, message) {
   // https://github.com/WebAssembly/spec/issues/840
-  assert_equals(Object.getPrototypeOf(actual), ArrayBuffer.prototype,
+  const bufferType = expected.shared ? SharedArrayBuffer : ArrayBuffer;
```
Comment (Ms2ger) [inline test/js-api/memory/grow.any.js]: Same question

File: test/js-api/memory/grow.any.js [hunk 5]
```diff
@@ -183,3 +184,29 @@ test(() => {
...
+  const memory = new WebAssembly.Memory(argument);
+  const oldMemory = memory.buffer;
+  assert_ArrayBuffer(oldMemory, { "size": 1, "shared": true }, "Buffer before growing");
+
+  const result = memory.grow(1);
+  assert_equals(result, 1);
+
+  const newMemory = memory.buffer;
+  assert_not_equals(oldMemory, newMemory);
+  assert_ArrayBuffer(oldMemory, { "size": 1, "shared": true }, "Old buffer after growing");
```
Comment (Ms2ger) [inline test/js-api/memory/grow.any.js]: Why is this not detached?

Comment (binji) [reply test/js-api/memory/grow.any.js]: SharedArrayBuffers can't be detached.

**Comment by binji** [comment]
> Looks like you still need to write the spec text for all of this?

Yes, I should probably have done that first. :-) But an old version of some of the changes is in the [overview](https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md#javascript-api-changes).

**Pull request merged**